### PR TITLE
Added section for AWS MachineSet spot instances

### PR DESCRIPTION
**Pull request by lbarbeevargas**
BZ-1860905: https://bugzilla.redhat.com/show_bug.cgi?id=1860905

@enxebre or @joelspeed PTAL. A few questions for SME review:

- Should this information also be included here: https://docs.openshift.com/container-platform/4.5/machine_management/creating-infrastructure-machinesets.html ?
- Does information regarding "a SpotMarketOptions needs to be added to the RunInstancesInput" need to be included in the customer-facing docs? https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/spot-instances.md#aws
- Is there any additional information that is missing?

### CODE REVIEW

File: modules/machineset-non-guaranteed-instance.adoc [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+//
+// * machine_management/creating-infrastructure-machinesets.adoc ?
+// * machine_management/creating_machinesets/creating-machineset-aws.adoc
+// * machine_management/creating_machinesets/creating-machineset-azure.adoc (4.6)
+// * machine_management/creating_machinesets/creating-machineset-gcp.adoc (4.6)
+
+[id="machineset-non-guaranteed-instance_{context}"]
+= Creating a spot instance
+
+You can create an AWS MachineSet that deploys machines as non-guaranteed, spot instances to save costs compared to on-demand instance prices. A machine is paired 1:1 with an instance on AWS. The MachineSet ensures the correct number of machines are created. If AWS does not have capacity, the Machine Health Checker (MHC) can remove the machine. The MHC also remediates a machine that fails if a request for a spot instance is not satisfied when a machine is created. AWS does not replace a spot instance when the instance is terminated.
```
Comment (JoelSpeed) [inline modules/machineset-non-guaranteed-instance.adoc:11]: > A machine is paired 1:1 with an instance on AWS. The MachineSet ensures the correct number of machines are created.

This is generic to MachineSets and not specific to spot instances so probably doesn't need saying here

I think we should have a brief explanation of what a spot instance is somewhere within the docs, ie, you get cheaper instances with the prerequisite agreement that AWS may terminate the instance at any point with a 2 minute warning (would also be good to explain that Openshift will react to this 2 minute warning by starting to drain the workloads from the affected instances)

File: modules/machineset-non-guaranteed-instance.adoc [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+.Procedure
+* Add the following fields under the `providerSpec` field in the MachineSet YAML file:
++
+[source,yaml]
+----
+providerSpec:
+  value:
+    spotMarketOptions: {}
+----
```
Comment (JoelSpeed) [inline modules/machineset-non-guaranteed-instance.adoc:26]: Should we add a note about the optional `maxPrice` field that can be used to limit the cost of the instance? If we do, we should also note that the default is to be charged up to the on-demand price and that it is strongly recommended not to set an explicit max price

**Review by JoelSpeed — COMMENTED** [review]
> Should this information also be included here: https://docs.openshift.com/container-platform/4.5/machine_management/creating-infrastructure-machinesets.html ?

I think it would be good to link from there at a minimum yes

> Does information regarding "a SpotMarketOptions needs to be added to the RunInstancesInput" need to be included in the customer-facing docs? https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/spot-instances.md#aws

This is more of an implementation detail than a user facing detail, so it should not be included

> Is there any additional information that is missing?

- Added a comment below about the optional `maxPrice` field
- We don't mention the termination notices or termination handlers, again, comment below
- Might be worth including some notes about the risks of running spot instances and the types of workloads or scenarios in which you may or may not want to run spot instances

**Comment by lbarbeevargas** [comment]
@JoelSpeed I applied your previous feedback. PTAL at new updates. Thank you!

File: modules/machineset-non-guaranteed-instance.adoc [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+//
+// * machine_management/creating-infrastructure-machinesets.adoc ?
+// * machine_management/creating_machinesets/creating-machineset-aws.adoc
+// * machine_management/creating_machinesets/creating-machineset-azure.adoc (4.6)
+// * machine_management/creating_machinesets/creating-machineset-gcp.adoc (4.6)
+
+[id="machineset-non-guaranteed-instance_{context}"]
+= MachineSet Spot Instances
+
+You can save costs by creating an AWS MachineSet that deploys machines as non-guaranteed, Spot Instances. Spot Instances utilize unused AWS EC2 capacity and are less expensive than On-Demand Instances. You can use Spot Instances for workloads that can tolerate interruptions, such as containerized and test and development workloads.
```
Comment (JoelSpeed) [inline modules/machineset-non-guaranteed-instance.adoc:11]:
```suggestion
You can save costs by creating an AWS MachineSet that deploys machines as non-guaranteed, Spot Instances. Spot Instances utilize unused AWS EC2 capacity and are less expensive than On-Demand Instances. You can use Spot Instances for workloads that can tolerate interruptions, such as batch or stateless, horizontally scalable workloads.
```

**Review by JoelSpeed — COMMENTED** [review]
Just one further suggestion, then LGTM

**Comment by lbarbeevargas** [comment]
> Just one further suggestion, then LGTM

Thanks for this suggestion @JoelSpeed. This was the part I was unsure about!

**Comment by lbarbeevargas** [comment]
@sunzhaohua2 PTAL. Thank you!

@openshift/team-documentation PTAL for master and 4.5. Thanks!

File: modules/machineset-non-guaranteed-instance.adoc [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+= MachineSet Spot Instances
+
+You can save costs by creating an AWS MachineSet that deploys machines as non-guaranteed, Spot Instances. Spot Instances utilize unused AWS EC2 capacity and are less expensive than On-Demand Instances. You can use Spot Instances for workloads that can tolerate interruptions, such as batch or stateless, horizontally scalable workloads.
+
+[IMPORTANT]
+====
+It is strongly recommended that control-plane machines are not created on Spot Instances due to the increased likelihood of the instance being terminated. Manual intervention is required to replace a terminated control-plane node.
+====
+
+AWS EC2 can terminate a Spot Instance at any time. AWS gives a two-minute warning to the user when an interruption occurs. The {product-title} begins to remove the workloads from the affected instances when AWS issues this termination notice.
```
Comment (codyhoag) [inline modules/machineset-non-guaranteed-instance.adoc:18]: Perhaps `this termination notice` could be changed to `the termination notice`? You described the warning and then when stating `this termination notice` directly afterwards, it made me initially think you were talking about the warning you just described. I assume that's not the case, and you're talking about after the two-minute warning expires as the termination period?

**Review by codyhoag — COMMENTED** [review]
Just a couple suggestions. Looks good otherwise! See Slack comment for additional info on suggested version PRs.

File: modules/machineset-non-guaranteed-instance.adoc (hunk 4 above)
Comment (lbarbeevargas) [reply modules/machineset-non-guaranteed-instance.adoc:18]: Good observation on the wording. I meant for `warning` and `notice` to be synonymous, but I can see how that is confusing as far as the timing of everything. My understanding is that OCP starts to remove the workloads when AWS issues the two-minute warning/notice. I can change `this termination notice` to `the termination warning` to make that more clear.

**Comment by lbarbeevargas** [comment]
I created [PR-24944](https://github.com/openshift/openshift-docs/pull/24944) to base these additions off of the `enterprise-4.5` branch instead of the `master` branch. Keeping this PR open until PR-24944 is merged.

File: machine_management/creating-infrastructure-machinesets.adoc [hunk 5]
```diff
@@ -37,6 +37,8 @@ Use the sample MachineSet for your cloud.
 
 include::modules/machineset-yaml-aws.adoc[leveloffset=+3]
 
+AWS MachineSets support non-guaranteed, xref:../machine_management/creating_machinesets/creating-machineset-aws.html#machineset-non-guaranteed-instance_creating-machineset-aws[Spot Instances], which can save you costs compared to On-Demand Instance prices. You can xref:../machine_management/creating_machinesets/creating-machineset-aws.html#machineset-creating-non-guaranteed-instance_creating-machineset-aws[configure Spot Instances] by adding SpotMarketOptions to the MachineSet YAML file.
```
Comment (codyhoag) [inline machine_management/creating-infrastructure-machinesets.adoc:40]: @lbarbeevargas the comma after `non-guaranteed` threw me off. Do you think it makes sense to remove that comma? Maybe I'm just reading it wrong. Your call :slightly_smiling_face:

**Comment by JoelSpeed** [comment]
All looks good from my perspective

**Comment by enxebre** [comment]
@lbarbeevargas are there analogous PRs for GCP and Azure?

**Comment by enxebre** [comment]
@lbarbeevargas 
>We don't mention the termination notices or termination handlers, again, comment below

This comments seems still not addressed. There's no mention to what happen on the machine API side when AWS decideds to terminate an instance, i.e the machine API will delete the machine resource, resulting in the creation of a new one to satisfy the machineSet replica number.

**Comment by lbarbeevargas** [comment]
> hey @lbarbeevargas are there analogous PRs for GCP and Azure?

Hi @enxebre, there will be new PRs for the 4.6 GCP and Azure work. I am starting on that content this sprint.

**Comment by lbarbeevargas** [comment]
> @lbarbeevargas
> 
> > We don't mention the termination notices or termination handlers, again, comment below
> 
> This comments seems still not addressed. There's no mention to what happen on the machine API side when AWS decideds to terminate an instance, i.e the machine API will delete the machine resource, resulting in the creation of a new one to satisfy the machineSet replica number.

@enxebre, thank you for following up on this. I can add this information in a new PR, as well as make sure it's in the 4.6 content for Azure and GCP.

**Comment by enxebre** [comment]
@lbarbeevargas 

>If AWS does not have capacity, the Machine Health Checker (MHC) can remove the machine.
The MHC also remediates a machine that fails if a request for a Spot Instance is not
satisfied when a machine is created. AWS does not replace a Spot Instance when the instance is terminated.

We plan to handle termination via deploying a MHC out of the box for spot instances but this is not implemented yet. I suggest we remove the statement above and add something like:
"When AWS terminates an instance a termination handler running on the spot instance node will delete the Machine resource. The machineSet will create a new one that will request a new spot instance"

**Pull request merged**